File: cf-mkv-extractor/src/ebml.js
// EBML / Matroska low-level primitives.
//
// Spec reference: https://www.rfc-editor.org/rfc/rfc8794 (EBML),
// https://www.matroska.org/technical/elements.html (Matroska element IDs).
//
// Two kinds of variable-length integers we care about:
//   * Element IDs       — keep the leading marker bit (it's part of the ID).
//   * Data sizes        — strip the leading marker bit (it's just length prefix).
// Both encode 1..8 bytes; the length is determined by the position of the
// first set bit in byte 0.

/**
 * Reads a variable-length integer starting at `buf[off]`.
 * @param {Uint8Array} buf
 * @param {number} off
 * @param {boolean} keepMarker  true → return the bits as-is (element ID),
 *                              false → strip the leading marker bit (data size).
 * @returns {{ value: number, length: number }}
 */
export function readVint(buf, off, keepMarker) {
    if (off >= buf.length) throw new Error('VINT: read past EOF');
    const first = buf[off];
    if (first === 0) throw new Error('VINT: invalid leading byte 0x00');

    let len = 1;
    let mask = 0x80;
    while (!(first & mask)) {
        mask >>>= 1;
        len++;
        if (len > 8) throw new Error('VINT: length > 8 bytes');
    }
    if (off + len > buf.length) {
        throw new Error('VINT: payload past EOF');
    }

    // JS bitwise ops cap at 32 bits, but Matroska sizes can be larger.
    // Use plain arithmetic for values that may exceed 2^32. We cap at
    // Number.MAX_SAFE_INTEGER which is ~9 PB — far above any real MKV.
    let value = keepMarker ? first : (first & (mask - 1));
    for (let i = 1; i < len; i++) {
        value = value * 256 + buf[off + i];
    }
    return { value, length: len };
}

/**
 * Reads an EBML element header: ID followed by data size.
 * Returns the element's ID, the byte offset where its DATA starts,
 * the data size, and the offset of the element AFTER this one.
 *
 * Handles the "unknown size" sentinel (all data bits set) by returning
 * `size === Infinity`. Callers walking such elements must stop at a
 * known terminating element instead of using `nextOffset`.
 */
export function readElement(buf, off) {
    const id = readVint(buf, off, true);
    const size = readVint(buf, off + id.length, false);

    // EBML unknown-size sentinel: all VINT data bits = 1. For an 8-byte
    // vint that's 0xFFFFFFFFFFFFFF (the marker bit isn't included in the
    // value). This is legal for Segment elements in live recordings.
    const isUnknownSize =
        size.length === 1 ? size.value === 0x7F :
        size.length === 2 ? size.value === 0x3FFF :
        size.length === 3 ? size.value === 0x1FFFFF :
        size.length === 4 ? size.value === 0x0FFFFFFF :
        size.length === 5 ? size.value === 0x07FFFFFFFF :
        size.length === 6 ? size.value === 0x03FFFFFFFFFF :
        size.length === 7 ? size.value === 0x01FFFFFFFFFFFF :
        size.length === 8 ? size.value === 0x00FFFFFFFFFFFFFF :
        false;

    const dataOffset = off + id.length + size.length;
    return {
        id: id.value,
        size: isUnknownSize ? Infinity : size.value,
        unknownSize: isUnknownSize,
        dataOffset,
        nextOffset: isUnknownSize ? Infinity : dataOffset + size.value,
    };
}

/** Reads an unsigned big-endian integer of 1..8 bytes. */
export function readUInt(buf, off, len) {
    let v = 0;
    for (let i = 0; i < len; i++) v = v * 256 + buf[off + i];
    return v;
}

/** Reads a signed big-endian integer (two's complement) of 1..8 bytes. */
export function readSInt(buf, off, len) {
    if (len === 0) return 0;
    let v = readUInt(buf, off, len);
    const sign = 1 << ((len * 8) - 1);
    if (v & sign) v -= 2 ** (len * 8);
    return v;
}

/** Reads a UTF-8 string of `len` bytes (trailing nulls stripped). */
export function readString(buf, off, len) {
    let end = off + len;
    while (end > off && buf[end - 1] === 0) end--;
    return new TextDecoder('utf-8', { fatal: false }).decode(buf.slice(off, end));
}

/** Returns the slice as a fresh Uint8Array (so callers can keep it around). */
export function readBytes(buf, off, len) {
    return buf.slice(off, off + len);
}

// Matroska element IDs we care about. Names match the spec for grep-ability.
export const ID = {
    EBML:                0x1A45DFA3,
    Segment:             0x18538067,

    SeekHead:            0x114D9B74,
    Seek:                0x4DBB,
    SeekID:              0x53AB,
    SeekPosition:        0x53AC,

    SegmentInfo:         0x1549A966,
    TimecodeScale:       0x2AD7B1,
    Duration:            0x4489,

    Tracks:              0x1654AE6B,
    TrackEntry:          0xAE,
    TrackNumber:         0xD7,
    TrackType:           0x83,
    CodecID:             0x86,
    CodecPrivate:        0x63A2,
    Name:                0x536E,
    Language:            0x22B59C,
    FlagDefault:         0x88,
    FlagForced:          0x55AA,

    Cluster:             0x1F43B675,
    ClusterTimecode:     0xE7,
    SimpleBlock:         0xA3,
    BlockGroup:          0xA0,
    Block:               0xA1,
    BlockDuration:       0x9B,

    Cues:                0x1C53BB6B,
    CuePoint:            0xBB,
    CueTime:             0xB3,
    CueTrackPositions:   0xB7,
    CueTrack:            0xF7,
    CueClusterPosition:  0xF1,
    CueRelativePosition: 0xF0,

    Void:                0xEC,
};

// Track types per spec — only the one we care about.
export const TRACK_TYPE_SUBTITLE = 0x11;

File: cf-mkv-extractor-rs/src/ebml.rs
//! EBML / Matroska low-level primitives.
//!
//! Port of cf-mkv-extractor/src/ebml.js. Same VINT semantics, same
//! element-ID constants. Where the JS version constructs short-lived
//! objects per element, this version returns plain structs by value
//! — no allocations in the hot path, which is the whole point of
//! moving to Rust for the cluster-walk loop.

#[derive(Debug, Clone, Copy)]
pub struct EbmlElement {
    pub id: u64,
    pub size: u64,
    pub data_offset: usize,
    pub next_offset: usize,
}

pub const TRACK_TYPE_SUBTITLE: u64 = 0x11;

// Element IDs we care about. Same values as ebml.js. Kept as u64
// because EBML IDs are VINTs with the marker bit retained.
pub mod id {
    pub const EBML_HEADER: u64 = 0x1A45_DFA3;
    pub const SEGMENT: u64 = 0x1853_8067;

    pub const SEEK_HEAD: u64 = 0x114D_9B74;
    pub const SEEK: u64 = 0x4DBB;
    pub const SEEK_ID: u64 = 0x53AB;
    pub const SEEK_POSITION: u64 = 0x53AC;

    pub const SEGMENT_INFO: u64 = 0x1549_A966;
    pub const TIMECODE_SCALE: u64 = 0x2AD7_B1;

    pub const TRACKS: u64 = 0x1654_AE6B;
    pub const TRACK_ENTRY: u64 = 0xAE;
    pub const TRACK_NUMBER: u64 = 0xD7;
    pub const TRACK_TYPE: u64 = 0x83;
    pub const CODEC_ID: u64 = 0x86;
    pub const CODEC_PRIVATE: u64 = 0x63A2;
    pub const NAME: u64 = 0x536E;
    pub const LANGUAGE: u64 = 0x22B5_9C;

    pub const CUES: u64 = 0x1C53_BB6B;
    pub const CUE_POINT: u64 = 0xBB;
    pub const CUE_TIME: u64 = 0xB3;
    pub const CUE_TRACK_POSITIONS: u64 = 0xB7;
    pub const CUE_TRACK: u64 = 0xF7;
    pub const CUE_CLUSTER_POSITION: u64 = 0xF1;

    pub const CLUSTER: u64 = 0x1F43_B675;
    pub const CLUSTER_TIMECODE: u64 = 0xE7;
    pub const SIMPLE_BLOCK: u64 = 0xA3;
    pub const BLOCK_GROUP: u64 = 0xA0;
    pub const BLOCK: u64 = 0xA1;
    pub const BLOCK_DURATION: u64 = 0x9B;
}

#[derive(Debug, Clone, Copy)]
pub struct Vint {
    pub value: u64,
    pub length: usize,
}

/// Reads a variable-length integer starting at buf[off].
/// keep_marker=true keeps the leading bit set (element IDs);
/// false strips it (data sizes).
pub fn read_vint(buf: &[u8], off: usize, keep_marker: bool) -> Result<Vint, &'static str> {
    if off >= buf.len() {
        return Err("VINT: read past EOF");
    }
    let first = buf[off];
    if first == 0 {
        return Err("VINT: invalid leading byte 0x00");
    }
    let mut len = 1usize;
    let mut mask = 0x80u8;
    while (first & mask) == 0 {
        mask >>= 1;
        len += 1;
        if len > 8 {
            return Err("VINT: length > 8 bytes");
        }
    }
    if off + len > buf.len() {
        return Err("VINT: payload past EOF");
    }
    let mut value: u64 = if keep_marker {
        first as u64
    } else {
        (first & !mask) as u64
    };
    for i in 1..len {
        value = (value << 8) | (buf[off + i] as u64);
    }
    Ok(Vint { value, length: len })
}

/// Reads an EBML element header at buf[off] — both ID and size VINTs.
pub fn read_element(buf: &[u8], off: usize) -> Result<EbmlElement, &'static str> {
    let id_v = read_vint(buf, off, true)?;
    let size_v = read_vint(buf, off + id_v.length, false)?;
    let data_offset = off + id_v.length + size_v.length;

    // EBML "unknown size" = all 1s after the marker. Treat as
    // max-u64 so callers can recognise and bail.
    let max_for_len = (1u64 << (7 * size_v.length as u32)) - 1;
    let size = if size_v.value == max_for_len {
        u64::MAX
    } else {
        size_v.value
    };
    let next_offset = if size == u64::MAX {
        usize::MAX
    } else {
        data_offset + size as usize
    };
    Ok(EbmlElement { id: id_v.value, size, data_offset, next_offset })
}

pub fn read_uint(buf: &[u8], off: usize, size: u64) -> u64 {
    let mut v = 0u64;
    let end = (off + size as usize).min(buf.len());
    for i in off..end {
        v = (v << 8) | (buf[i] as u64);
    }
    v
}

pub fn read_string(buf: &[u8], off: usize, size: u64) -> String {
    let end = (off + size as usize).min(buf.len());
    let slice = &buf[off..end];
    let trimmed = slice
        .iter()
        .position(|&b| b == 0)
        .map(|i| &slice[..i])
        .unwrap_or(slice);
    String::from_utf8_lossy(trimmed).into_owned()
}

pub fn read_bytes(buf: &[u8], off: usize, size: u64) -> Vec<u8> {
    let end = (off + size as usize).min(buf.len());
    buf[off..end].to_vec()
}

/// Read a signed 16-bit big-endian integer at buf[off].
/// (SimpleBlock timecode field.)
pub fn read_i16_be(buf: &[u8], off: usize) -> i16 {
    if off + 2 > buf.len() {
        return 0;
    }
    ((buf[off] as i16) << 8) | (buf[off + 1] as i16)
}

/// Walks a contiguous EBML buffer scanning for a direct-child
/// element with the requested ID. Returns the element header,
/// or None if not found within the bounds.
pub fn find_child(buf: &[u8], start: usize, end: usize, target_id: u64) -> Option<EbmlElement> {
    let mut off = start;
    while off < end && off < buf.len() {
        let el = read_element(buf, off).ok()?;
        if el.id == target_id {
            return Some(el);
        }
        if el.size == u64::MAX {
            return None;
        }
        off = el.next_offset;
        if off == usize::MAX {
            return None;
        }
    }
    None
}
